src/lightning/data/dataset.py

```diff
@@ -0,0 +1,89 @@
+import os
+import tempfile
+from abc import ABC, abstractmethod
+from typing import Any, Dict, Optional, Tuple
+
+from torch.utils.data import Dataset as TorchDataset
+
+from lightning.data.authenticators import _DatasetBackend, LocalDatasetBackend, S3DatasetBackend
+from lightning.data.dataset_index import get_index
+from lightning.data.fileio import OpenCloudFileObj
+
+
+class LightningDataset(TorchDataset, ABC):
+    """Dataset wrapper for optimized dataloading.
+
+    Arguments:
+
+        data_source: path of data directory.
```

Comment:
Maybe explain whether this incorporates the prefix in case of S3, i.e. `s3://`

Reply:
yes, we'll have examples for this!